hera_cal/vis_clean.py

```diff
@@ -474,20 +498,214 @@ def fft(d, ifft=ifft, shift=fftshift, ax=-1):
+            such that the windowing function smoothly approaches zero. If axis is tuple,
+            can feed as a tuple specifying for each FFT axis.
+        ifft : bool, if True, use ifft instead of fft
+        ifftshift : bool, if True, ifftshift data along FT axis before FFT.
+        fftshift : bool, if True, fftshift along FT axes after FFT.
+        zeropad : int, number of zero-valued channels to append to each side of FFT axis.
+            If axis is tuple, can feed as a tuple specifying for each FFT axis.
+    Returns:
+        dfft : complex ndarray FFT of data
+        fourier_axes : fourier axes, if axis is ndimensional, so is this.
+    """
+    # type checks
+    if not isinstance(axis, (tuple, list)):
+        axis = [axis]
+    if not isinstance(window, (tuple, list)):
+        window = [window for i in range(len(axis))]
+    if not isinstance(alpha, (tuple, list)):
+        alpha = [alpha for i in range(len(axis))]
+    if not isinstance(edgecut_low, (tuple, list)):
+        edgecut_low = [edgecut_low for i in range(len(axis))]
+    if not isinstance(edgecut_hi, (tuple, list)):
+        edgecut_hi = [edgecut_hi for i in range(len(axis))]
+    if not isinstance(zeropad, (tuple, list)):
+        zeropad = [zeropad for i in range(len(axis))]
+    if not isinstance(delta_bin, (tuple, list)):
+        if len(axis) > 1:
+            raise ValueError("delta_bin must have same len as axis")
+        delta_bin = [delta_bin]
+    else:
+        if len(delta_bin) != len(axis):
+            raise ValueError("delta_bin must have same len as axis")
+    Nax = len(axis)
+
+    # get a copy
+    data = data.copy()
+
+    # set fft convention
+    fourier_axes = []
+    if ifft:
+        fft = np.fft.ifft
+    else:
+        fft = np.fft.fft
+
+    # get wgts
+    if wgts is None:
+        wgts = np.ones_like(data, dtype=np.float)
+    data *= wgts
+
+    # iterate over axis
+    for i, ax in enumerate(axis):
+        Nbins = data.shape[ax]
+
+        # generate and apply window
+        win = dspec.gen_window(window[i], Nbins, alpha=alpha[i], edgecut_low=edgecut_low[i], edgecut_hi=edgecut_hi[i])
+        wshape = np.ones(data.ndim, dtype=np.int)
+        wshape[ax] = Nbins
+        win.shape = tuple(wshape)
+        data *= win
+
+        # zeropad
+        data = _zeropad_array(data, zeropad[i], axis=ax)
+
+        # ifftshift
+        if ifftshift:
+            data = np.fft.ifftshift(data, axes=ax)
+
+        # FFT
+        data = fft(data, axis=ax)
+
+        # get fourier axis
+        fax = np.fft.fftfreq(data.shape[ax], delta_bin[i])
+
+        # fftshift
+        if fftshift:
+            data = np.fft.fftshift(data, axes=ax)
+            fax = np.fft.fftshift(fax)
+
+        fourier_axes.append(fax)
+
+    if len(axis) == 1:
+        fourier_axes = fourier_axes[0]
+
+    return data, fourier_axes
+
+
+def _zeropad_array(data, zeropad=0, axis=-1, undo=False):
```

Comment:
I think this function is reinventing the wheel. Numpy already has zeropadding in `np.pad` which covers most of the functionality. I can see an argument for writing a function that undoes the effect of np.pad with a set of kwargs, but we should stick to standard libraries and syntax as much as possible.

Reply:
Ah I didn't know the function existed. Regardless, this function seems to have two critical capabilities that `np.pad` doesn't: 1) support for an `axis` and 2) and `undo` kwarg.

I also don't think not using `np.pad` in this case is that big of a deal, honestly...